pkgs/development/python-modules/playwright/default.nix

```diff
@@ -0,0 +1,94 @@
+{ lib
+, buildPythonPackage
+, fetchFromGitHub
+, pythonOlder
+, python
+, greenlet
+, websockets
+, pyee
+, pillow
+, pixelmatch
+, autobahn
+, pyopenssl
+, service-identity
+, requests
+, objgraph
+, callPackage
+, nodejs
+}:
+
+buildPythonPackage rec {
+  pname = "playwright";
+  version = "1.22.0";
+  disabled = pythonOlder "3.7";
+
+  src = fetchFromGitHub {
+    owner = "microsoft";
+    repo = "playwright-python";
+    rev = "v${version}";
+    sha256 = "sha256-nodm1Wts61vRCqjqNX54tD0cQB7uoxx6Eypz7nBR54E=";
+  };
+
+  format = "pyproject";
+
+  playwright-driver = callPackage ./driver.nix { playwrightVersion = version; };
+
+  postPatch = ''
+    # works with pyee 9 as well, which is what nixos has packaged
+    # disables setuptools-scm, as it fails
+    substituteInPlace meta.yaml --replace "- pyee ==8.1.0" ""
+    substituteInPlace setup.py \
+        --replace '"pyee==8.1.0",' "" \
+        --replace '"setuptools-scm==6.3.2",' ""
+    # replace functionality that setuptools-scm provides by adding the version to a file
+    echo 'version = "${version}"' > playwright/_repo_version.py
+  '';
+
+  preBuild = ''
+    mkdir -p driver
+    cp -r ${playwright-driver}/driver ./
+  '';
+
+  preInstall = ''
+    # pip names the wheel "dist", but when installing it will
+    # expect a properly formatted filename in dist.
+    mv dist dist.orig
+    mkdir -p dist
+    mv dist.orig dist/playwright-${version}-py3-none-any.whl
+  '';
+
+  propagatedBuildInputs = [
+    playwright-driver
+    greenlet
+    websockets
+    pyee
+    pillow
+    pixelmatch
+    autobahn
+    pyopenssl
+    service-identity
+    requests
+    objgraph
+  ];
+
+  # skip tests because they require network access
+  doCheck = false;
+
+  pythonImportsCheck = [
+    "playwright"
+  ];
+
+  postInstall = ''
+    # use nix's nodejs instead of playwright's nodejs
+    substituteInPlace "$out/lib/${python.libPrefix}/site-packages/playwright/driver/playwright.sh" \
+      --replace '"$SCRIPT_PATH/node"' '"${nodejs}/bin/node"'
```

Comment:
This should actually make `playwright.sh` runnable:

```suggestion
    pushd $out/lib/${python.libPrefix}/site-packages/playwright/driver

    # use nix's nodejs instead of playwright's nodejs
    substituteInPlace playwright.sh --replace '"$SCRIPT_PATH/node"' '"${nodejs}/bin/node"'
    rm node

    # Hard-code the script path to be our $out directory. The alternative would be to add coreutils
    # to the closure and use realpath like the upstream script does.
    substituteInPlace playwright.sh --replace \
      'SCRIPT_PATH="$(cd "$(dirname "$0")" ; pwd -P)"' \
      "SCRIPT_PATH=$(pwd)"

    patchShebangs playwright.sh
    popd
```

The interesting (new) stuff here is patching the shebang (upstream uses `/bin/sh`) and replacing the SCRIPT_PATH with a static pointer to `$out`. The alternative would have been to add coreutils to the closure, and leave the script as-is, but this should work as well.

Funny thing is, by now we have patched every single line from the upstream script. It might even be easier to do something like this instead:

```bash
pushd "$out/lib/${python.libPrefix}/site-packages/playwright/driver"
rm playwright.sh node
makeWrapper ${nodejs}/bin/node playwright.sh \
  --add-flags "$(pwd)/package/lib/cli/cli.js"
popd
```

Not sure which of the two approaches would be better maintainable in the long term. I have no idea where `playwright.sh` comes from and how often it's changed.